vtex-apps/binding-selector
# Feat/keep sales channel

#### What problem is this solving?


For some use cases, it's important to keep the user in the same sales channel when changing binding, e.g. **AmetllerOrigen**. The current behavior for the app is to change the sales channel back to the main one when user changes it, even when the flag `Update Sales Channel` is off in the admin settings.

This PR changes that behavior, keeping the user always in the first sales channel they navigated to when the flag `Update Sales Channel` is off.

To accomplish that, the first time user changes the binding, the app grabs the sales channel on `window.__RUNTIME__.segmentToken` and adds a query string in the URL the user will be redirected, e.g. `sc=1`. When the URL has that query string, the app uses the query string to save the target sales channel into state. This was needed because after the first load, `window.__RUNTIME__.segmentToken` gets out of sync. If user changes binding again, now the app uses the sales channel set in the state.

This approach works because the `vtex_session` is set for paths considering the binding, e.g. `/es` or `/pt`. It makes the app even more dependent of `binding-countries` app to avoid users hitting the `/` path.

#### How to test it?


The branch is linked in this [workspace](https://bskeepsc--powerplanet.myvtex.com) but it's also installed on [master](https://www.testperformancev.com).

To test it:

1. Clean the cookies;
2. Visit any binding from the navigation bar;
3. Change binding;
4. Add something to cart and go to checkout page.

Checking the sales channel in the order form - `vtexjs.checkout.orderForm.salesChannel`, it should match with the one related to the first binding you've visited. (es:1, pt:5, fr:4).

Try to repeat the test making the first visit from a different binding (clean the cookies again!).

#### Will it work on Ametller?

Hard to tell, because the cookies behavior is quite erratic, and difficult to predict how it will work with a different app also trying make changes in the session.

That's been said, I think the app is gonna work this way: after user selects a zipcode, Ametller patches the new sales channel in the session, then reloads the page. Since it all happens in the same path, the `window.__RUNTIME__.segmentToken` will be updated as well. If the user changes the binding using `binding-selector`, the app will grab that value as the first sales channel visited.

The behavior of the cookies is quite of erratic, and any thoughts on edge cases are very well appreciated.
